File: src/trading/DeriveExchange.ts
import { DeriveOptionsTrader } from "./derive/derive";
import { OptionsExchange } from "./OptionExchange";

export class DeriveExchange implements OptionsExchange {
  name = "Derive";
  private trader: DeriveOptionsTrader;

  constructor() {
    this.trader = new DeriveOptionsTrader(
      process.env.Derive_CLIENT_ID!,
      process.env.Derive_CLIENT_SECRET!,
      true
    );
  }

  async placeOrder(params: {
    instrument_name: string;
    side: "buy" | "sell";
    price: number;
    quantity: number;
    order_type?: "limit" | "market";
    time_in_force?: "fill_or_kill";
  }) {
    return this.trader.placeOrder({
      instrument_name: params.instrument_name,
      side: params.side,
      amount: params.quantity,
      price: params.price,
      order_type: params.order_type ?? "limit",
      time_in_force: params.time_in_force ?? "fill_or_kill",
    });
  }
}

File: src_python/trading/derive_exchange.py
from __future__ import annotations

import os

from src_python.trading.derive.derive import DeriveOptionsTrader


class DeriveExchange:
    name = "derive"

    def __init__(self) -> None:
        self.trader = DeriveOptionsTrader(
            os.getenv("Derive_CLIENT_ID", ""),
            os.getenv("Derive_CLIENT_SECRET", ""),
            True,
        )

    async def place_order(
        self,
        *,
        instrument_name: str,
        side: str,
        price: float,
        quantity: float,
        underlying_price: float | None = None,
        order_type: str = "limit",
        time_in_force: str = "fill_or_kill",
    ) -> object:
        return await self.trader.place_order(
            instrument_name=instrument_name,
            side=side,
            amount=quantity,
            price=price,
            order_type=order_type,
            time_in_force=time_in_force,
        )
